samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -0,0 +1,166 @@
+import org.I0Itec.zkclient.IZkDataListener;
+import org.apache.samza.SamzaException;
+import org.apache.samza.coordinator.leaderelection.LeaderElector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * <p>
+ * An implementation of Leader Elector using Zookeeper.
+ *
+ * Each participant in the leader election process creates an instance of this class and tries to become the leader.
+ * The participant with the lowest sequence number in the ZK subtree for election becomes the leader. Every non-leader
+ * sets a watcher on its predecessor, where the predecessor is the participant with the largest sequence number
+ * that is less than the current participant's sequence number.
+ * </p>
+ * */
+public class ZkLeaderElector implements LeaderElector {
+  public static final Logger LOGGER = LoggerFactory.getLogger(ZkLeaderElector.class);
+  private final ZkUtils zkUtils;
+  private final String processorIdStr;
+  private final ZkKeyBuilder keyBuilder;
+  private final String hostName;
+
+  private String leaderId = null;
+  private final IZkDataListener zkLeaderElectionListener;
+  private String currentSubscription = null;
+  private final Random random = new Random();
+
+  @VisibleForTesting
+  ZkLeaderElector(String processorIdStr, ZkUtils zkUtils, IZkDataListener leaderElectionListener) {
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.zkLeaderElectionListener = leaderElectionListener;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  public ZkLeaderElector(String processorIdStr, ZkUtils zkUtils) {
+    this.zkLeaderElectionListener = new ZkLeaderElectionListener();
+    this.processorIdStr = processorIdStr;
+    this.zkUtils = zkUtils;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    this.hostName = getHostName();
+  }
+
+  // TODO: This should go away once we integrate with Zk based Job Coordinator
+  private String getHostName() {
+    try {
+      return InetAddress.getLocalHost().getHostName();
+    } catch (UnknownHostException e) {
+      LOGGER.error("Failed to fetch hostname of the processor", e);
+      throw new SamzaException(e);
+    }
+  }
+
+  @Override
+  public boolean tryBecomeLeader() {
+    String currentPath = zkUtils.getEphemeralPath();
+
+    if (currentPath == null || currentPath.isEmpty()) {
+      currentPath = zkUtils.registerProcessorAndGetId(hostName);
+    }
+
+    List<String> children = zkUtils.getActiveProcessors();
+    LOGGER.debug(zLog("Current active processors - " + children));
+    int index = children.indexOf(ZkKeyBuilder.parseIdFromPath(currentPath));
+
+    if (children.size() == 0 || index == -1) {
+      throw new SamzaException("Looks like we are no longer connected to Zk. Need to reconnect!");
+    }
+
+    leaderId = ZkKeyBuilder.parseIdFromPath(children.get(0));
```

Comment:
Seems the leaderId can be expired after the initial phase. If the leader is dead in the future and the current processor might not be aware of it. I think it might be better not caching the leadId here. If this is only needed for amLeader() function, it would be better to keep a boolean flag.